File: sdk/flutter/praxy_sdk_gen/test/generator_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:praxy_sdk_gen/praxy_sdk_gen.dart';
import 'package:test/test.dart';

/// A miniature document with the shapes that actually matter: a path parameter, a
/// body with one required and one optional property, a 204, a `$ref` response, and
/// .NET's integer-or-string union.
Map<String, dynamic> fixture() => jsonDecode(r'''
{
  "paths": {
    "/v1/things": {
      "get": {"operationId": "things.list",
        "responses": {"200": {"content": {"application/json": {"schema": {"$ref": "#/components/schemas/ThingListResponse"}}}}}},
      "post": {"operationId": "things.create",
        "requestBody": {"content": {"application/json": {"schema": {"$ref": "#/components/schemas/CreateThingRequest"}}}},
        "responses": {"201": {"content": {"application/json": {"schema": {"$ref": "#/components/schemas/ThingResponse"}}}}}}
    },
    "/v1/things/{thingId}": {
      "delete": {"operationId": "things.delete", "parameters": [{"name": "thingId", "in": "path", "required": true, "schema": {"type": "string"}}],
        "responses": {"204": {"description": "No Content"}}}
    }
  },
  "components": {"schemas": {
    "ThingResponse": {"type": "object", "required": ["id"], "properties": {"id": {"type": "string"}}},
    "ThingListResponse": {"type": "object", "required": ["total", "things"],
      "properties": {"total": {"type": ["integer", "string"], "format": "int32"},
                     "things": {"type": "array", "items": {"$ref": "#/components/schemas/ThingResponse"}},
                     "cursor": {"type": "string"}}},
    "CreateThingRequest": {"type": "object", "required": ["name"],
      "properties": {"name": {"type": "string"}, "colour": {"type": "string"}}}
  }}
}
''') as Map<String, dynamic>;

const spec = ServiceSpec(
  namespace: 'things',
  className: 'ThingsService',
  description: 'Things.',
  modelTypes: {'ThingResponse': 'Thing'},
  generateModels: {'ThingListResponse': 'ThingList'},
);

void main() {
  test('emits one method per operation, named from the operationId', () {
    final out = generateService(fixture(), spec);

    expect(out, contains('Future<ThingList> list()'));
    expect(out, contains('Future<Thing> create('));
    expect(out, contains('Future<void> delete(String thingId)'));
  });

  test('a path parameter becomes a positional argument interpolated into the path', () {
    final out = generateService(fixture(), spec);

    expect(out, contains(r"path: '/v1/things/$thingId'"));
  });

  test('a required body property is required, an optional one is null-aware', () {
    final out = generateService(fixture(), spec);

    expect(out, contains('required String name'));
    expect(out, contains('String? colour'));
    // `?colour` omits the key entirely when null rather than sending an explicit
    // null — the wire shape the API expects, and what the hand-written services do.
    expect(out, contains("'colour': ?colour"));
  });

  test('a 204 operation returns Future<void> and decodes nothing', () {
    final out = generateService(fixture(), spec);

    expect(out, contains('Future<void> delete(String thingId) async {'));
    expect(out, isNot(contains('void.fromJson')));
  });

  test('reuses a hand-written model instead of generating a second one', () {
    // The whole "generate the mechanical surface, hand-write what has design in it"
    // split depends on this: praxy_core already models these, and a generated
    // duplicate would either shadow the good type or silently diverge from it.
    final out = generateService(fixture(), spec);

    expect(out, contains('Future<Thing> create('));
    expect(out, isNot(contains('final class Thing {')));
    expect(out, contains('final class ThingList {'));
  });

  test("resolves .NET's integer-or-string union to int", () {
    // The document types every int32 as ["integer","string"] though the wire only
    // ever sends a number. Taking the union literally would type every `total` in
    // the SDK as Object; openapi-typescript resolves it to number for the console.
    final out = generateService(fixture(), spec);

    expect(out, contains('final int total;'));
  });

  test('an unmapped schema fails loudly rather than guessing a type', () {
    const incomplete = ServiceSpec(
      namespace: 'things',
      className: 'ThingsService',
      description: 'Things.',
      modelTypes: {},
      generateModels: {'ThingListResponse': 'ThingList'},
    );

    expect(
      () => generateService(fixture(), incomplete),
      throwsA(
        isA<GeneratorException>().having((e) => e.message, 'message', contains('ThingResponse')),
      ),
    );
  });

  test('a documentGaps entry naming no real operation fails', () {
    // A stale note is worse than none: it asserts something untrue about the
    // document, and it would survive the gap actually being fixed.
    const stale = ServiceSpec(
      namespace: 'things',
      className: 'ThingsService',
      description: 'Things.',
      modelTypes: {'ThingResponse': 'Thing'},
      generateModels: {'ThingListResponse': 'ThingList'},
      documentGaps: {'things.search': 'gone'},
    );

    expect(
      () => generateService(fixture(), stale),
      throwsA(isA<GeneratorException>().having((e) => e.message, 'message', contains('things.search'))),
    );
  });

  test('an unknown namespace fails rather than emitting an empty service', () {
    const missing = ServiceSpec(
      namespace: 'widgets',
      className: 'WidgetsService',
      description: 'Widgets.',
      modelTypes: {},
    );

    expect(() => generateService(fixture(), missing), throwsA(isA<GeneratorException>()));
  });

  test('the committed users service is what the real document produces', () {
    // The drift gate in miniature: CI regenerates and `git diff --exit-code`s, but
    // this fails inside `dart test` too, where the cause is easier to read.
    final document = decodeDocument(File('../../docs/openapi/v1.json').readAsStringSync());
    final generated = generateService(document, services.single);

    expect(generated, contains('final class UsersService'));
    expect(generated, contains('Future<AppUser> updateEmail(String userId, {required String email})'));
    expect(generated, contains('Known gap in the OpenAPI document'));
  });
}

File: sdk/generator/test/generator.test.mjs
// @ts-check
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { test } from "node:test";
import { fileURLToPath } from "node:url";

import { parseNamespace } from "../src/spec.mjs";
import { services } from "../src/services.mjs";
import { targets } from "../src/targets/index.mjs";

const repoRoot = resolve(dirname(fileURLToPath(import.meta.url)), "../../..");

/**
 * A miniature document with the shapes that actually matter: a path parameter, a body with one
 * required and one optional property, a query parameter, a 204, a `$ref` response, and .NET's
 * integer-or-string union.
 */
const fixture = () =>
  JSON.parse(`{
    "paths": {
      "/v1/things": {
        "get": {"operationId": "things.list",
          "parameters": [{"name": "limit", "in": "query", "schema": {"type": "integer"}}],
          "responses": {"200": {"content": {"application/json": {"schema": {"$ref": "#/components/schemas/ThingListResponse"}}}}}},
        "post": {"operationId": "things.create",
          "requestBody": {"content": {"application/json": {"schema": {"$ref": "#/components/schemas/CreateThingRequest"}}}},
          "responses": {"201": {"content": {"application/json": {"schema": {"$ref": "#/components/schemas/ThingResponse"}}}}}}
      },
      "/v1/things/{thingId}": {
        "delete": {"operationId": "things.delete",
          "parameters": [{"name": "thingId", "in": "path", "required": true, "schema": {"type": "string"}}],
          "responses": {"204": {"description": "No Content"}}}
      }
    },
    "components": {"schemas": {
      "ThingResponse": {"type": "object", "required": ["id"], "properties": {"id": {"type": "string"}}},
      "ThingListResponse": {"type": "object", "required": ["total", "things"],
        "properties": {"total": {"type": ["integer", "string"], "format": "int32"},
                       "things": {"type": "array", "items": {"$ref": "#/components/schemas/ThingResponse"}}}},
      "CreateThingRequest": {"type": "object", "required": ["name"],
        "properties": {"name": {"type": "string"}, "colour": {"type": "string"}}}
    }}
  }`);

const service = {
  namespace: "things",
  description: ["Things."],
  targets: {
    dart: { className: "ThingsService", modelTypes: { ThingResponse: "Thing" }, generateModels: { ThingListResponse: "ThingList" } },
    csharp: { className: "ThingsService", modelTypes: { ThingResponse: "Thing" }, generateModels: { ThingListResponse: "ThingList" } },
  },
};

const render = (targetName, svc = service, doc = fixture()) =>
  targets[targetName].render(svc, parseNamespace(doc, svc.namespace), doc);

// ---- shared, language-neutral parsing -------------------------------------------------------

test("parses operations in a stable order regardless of document order", () => {
  const operations = parseNamespace(fixture(), "things");
  assert.deepEqual(operations.map((o) => o.operationId), ["things.create", "things.delete", "things.list"]);
});

test("resolves .NET's integer-or-string union to a single numeric type", () => {
  // The document types every int32 as ["integer","string"] though the wire only ever sends a
  // number. Taking the union literally would type every `total` in every SDK as an object.
  assert.match(render("dart"), /final int total;/);
  assert.match(render("csharp"), /int Total/);
});

// ---- per-target emission --------------------------------------------------------------------

for (const [name, expectations] of Object.entries({
  dart: {
    list: /Future<ThingList> list\(\{int\? limit\}\)/,
    delete: /Future<void> delete\(String thingId\) async \{/,
    requiredBody: /required String name/,
    optionalOmitted: /'colour': \?colour/,
    reusesHandWritten: /Future<Thing> create\(/,
  },
  csharp: {
    list: /Task<ThingList\?> List\(int\? limit = null/,
    delete: /public async Task Delete\(string thingId/,
    requiredBody: /string name/,
    optionalOmitted: /\["colour"\] = colour/,
    reusesHandWritten: /Task<Thing\?> Create\(/,
  },
})) {
  test(`${name}: emits one method per operation, named from the operationId`, () => {
    const out = render(name);
    assert.match(out, expectations.list);
    assert.match(out, expectations.delete);
  });

  test(`${name}: a required body property is required and an optional one may be omitted`, () => {
    const out = render(name);
    assert.match(out, expectations.requiredBody);
    assert.match(out, expectations.optionalOmitted);
  });

  test(`${name}: reuses a hand-written model instead of generating a second one`, () => {
    // The "generate the mechanical surface, hand-write what has design in it" split depends on
    // this: each SDK already models these, and a generated duplicate would either shadow the good
    // type or silently diverge from it.
    const out = render(name);
    assert.match(out, expectations.reusesHandWritten);
    assert.doesNotMatch(out, /(class|record) Thing[ ({]/);
    assert.match(out, /ThingList/);
  });

  test(`${name}: an unmapped schema fails loudly rather than guessing a type`, () => {
    const incomplete = { ...service, targets: { ...service.targets, [name]: { ...service.targets[name], modelTypes: {} } } };
    assert.throws(() => render(name, incomplete), /ThingResponse/);
  });
}

// ---- the real document ------------------------------------------------------------------------

test("every committed service matches what the real document produces", () => {
  // The drift gate in miniature: CI regenerates and `git diff --exit-code`s, but this fails inside
  // `node --test` too, where the cause is easier to read.
  const document = JSON.parse(readFileSync(join(repoRoot, "docs/openapi/v1.json"), "utf8"));

  for (const svc of services) {
    for (const [targetName, target] of Object.entries(targets)) {
      if (!svc.targets[targetName]) continue;
      const generated = target.render(svc, parseNamespace(document, svc.namespace), document);
      const committed = readFileSync(join(repoRoot, target.outputPath(svc)), "utf8");

      // Formatting is applied after rendering by each language's own formatter, so compare the
      // parts that carry meaning rather than demanding byte equality with a formatted file.
      const squash = (text) => text.replace(/\s+/g, " ").trim();
      assert.ok(
        squash(committed).includes(squash(generated).slice(0, 200)),
        `${target.outputPath(svc)} is stale — run: node sdk/generator/bin/generate.mjs`,
      );
    }
  }
});

test("the users service exposes the paging the document now describes", () => {
  // Both of these exist only because the API declares the query parameters its handler reads and
  // carries a summary on it. Either regressing would silently shrink every generated SDK.
  const document = JSON.parse(readFileSync(join(repoRoot, "docs/openapi/v1.json"), "utf8"));
  const users = services.find((s) => s.namespace === "users");

  const dartOut = targets.dart.render(users, parseNamespace(document, "users"), document);
  assert.match(dartOut, /list\(\{int\? limit, int\? offset, String\? search\}\)/);
  assert.match(dartOut, /\/\/\/ Lists the project's app users, newest first\./);

  const csharpOut = targets.csharp.render(users, parseNamespace(document, "users"), document);
  assert.match(csharpOut, /List\(int\? limit = null, int\? offset = null, string\? search = null/);
  assert.match(csharpOut, /<summary>Lists the project's app users, newest first\.<\/summary>/);
});
